documentation/specs/project-cache.md

```diff
@@ -14,106 +14,195 @@ This change also simplifies and unifies user experiences. MSBuild works the same
+      - Potential Issue: serialization format may change between writing and reading the `BuildResult`, especially if binary serialization is used.
+
+## Cache add scenario
+
+- Upon a cache miss, MSBuild will generally handle a request as normal, ie by building it.
+- To facilitate the plugin being able to hande future builds, MSBuild can report file accesses as well as the build result for the BuildRequestData.
+- MSBuild disables the in-proc node and uses [Detours](https://github.com/microsoft/Detours) to observe file accesses of the worker nodes. It forwards this information to the plugin for it to use as desired.
+  - This functionality has some implementation restrictions so will require additional opt-in. Specifically, the `/ReportFileAccesses` command-line flag or by setting `BuildParameters.ReportFileAccesses` for programatic use of `BuildManager`. If this is not set, no file accesses will be reported to the plugin, however the plugin will still be notified of the build result.
+- Due to the experimental nature of the feature, `/ReportFileAccesses` is only available with MSBuild.exe (ie. the Visual Studio install; not `dotnet`) and only from the command-line. The Visual Studio IDE does not set `BuildParameters.ReportFileAccesses`.
+- As described above, it is recommended to serialize the `BuildResult` from `HandleProjectFinishedAsync` for later replay.
+
+# APIs and calling patterns
+
+## Plugin API
+[ProjectCachePluginBase](/src/Build/BackEnd/Components/ProjectCache/ProjectCachePluginBase.cs) is an abstract class which plugin implementors will subclass. It contains the following methods:
+
+```cs
+public abstract Task BeginBuildAsync(CacheContext context, PluginLoggerBase logger, CancellationToken cancellationToken);
+
+public abstract Task EndBuildAsync(PluginLoggerBase logger, CancellationToken cancellationToken);
+
+public abstract Task<CacheResult> GetCacheResultAsync(BuildRequestData buildRequest, PluginLoggerBase logger, CancellationToken cancellationToken);
+
+public virtual void HandleFileAccess(FileAccessContext fileAccessContext, FileAccessData fileAccessData);
+
+public virtual void HandleProcess(FileAccessContext fileAccessContext, ProcessData processData);
+
+public virtual Task HandleProjectFinishedAsync(FileAccessContext fileAccessContext, BuildResult buildResult, PluginLoggerBase logger, CancellationToken cancellationToken);
+```
```

Comment:
This feels superfluous to the link to iplementation provided

Reply:
Also it should probably link the `Deteours` section below (and leave the implementation details info - like the need to disable in-proc node etc. - to be mentioned only in that section. Now some info is duplicated)

Reply:
Removing the signatures and linked to the `Plugin implementation guidance and simple example design` section.